File: BoardGameTracker.Common/Models/GameStatistics.cs
namespace BoardGameTracker.Common.Models;

public class GameStatistics
{
    public int PlayCount { get; set; }
    public double TotalPlayedTime { get; set; }
    public double? PricePerPlay { get; set; }
    public double? HighScore { get; set; }
    public double? AverageScore { get; set; }
    public MostWinningPlayer? MostWinsPlayer { get; set; }
    public double AveragePlayTime { get; set; }
    public DateTime? LastPlayed { get; set; }
    public int? ExpansionCount { get; set; }
}

public class MostWinningPlayer
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Image { get; set; } = string.Empty;
    public int TotalWins { get; set; }
}

File: boardgametracker.client/src/models/Games/GameStatistics.ts
import { Player } from '..';

export interface MostWinner extends Player {
  totalWins: number;
}

export interface GameStatistics {
  gameStats: GameStats;
  topPlayers: TopPlayer[];
  playByDayChart: PlayByDayChartData[];
  playerCountChart: PlayerCountChartData[];
  playerScoringChart: PlayerScoringChartData[];
  scoreRankChart: ScoreRankChartData[];
}

export interface GameStats {
  playCount: number;
  totalPlayedTime: number;
  pricePerPlay: number | null;
  highScore: number | null;
  averageScore: number | null;
  averagePlayTime: number | null;
  mostWinsPlayer: MostWinner | null;
  lastPlayed: string | null;
  expansionCount: number | null;
}

export interface TopPlayer {
  playerId: number;
  playCount: number;
  wins: number;
  winPercentage: number;
  trend: Trend;
}

export interface PlayByDayChartData {
  dayOfWeek: number;
  playCount: number;
}

export interface PlayerCountChartData {
  players: number;
  playCount: number;
}

export interface PlayerScoringChartData {
  dateTime: Date | string;
  series: XValue[];
}

export interface ScoreRankChartData {
  key: string;
  score: number;
  playerId: number;
}

export interface XValue {
  id: string;
  value: number;
}

export enum Trend {
  Up = 0,
  Down = 1,
  Equal = 2,
}
